### PR TITLE
Implement allowance ranges 2

### LINKED ISSUE #603: Implement MARECO ranges
When applying a standard allowance on a train run, it will automatically go from the beginning of the path to the end, with the value the user asked for.

Depending on the type of path and/or the rolling stock, the value of the margin might differ on some ranges (for example 5% for 10km, then 4.5mn/100km for 200km, then again 5%, covering the rest of the path).
All combined, theses ranges with different margins cover the entire path.

The user must be able to ask for these specified ranges and the core must compute a train run, allocating the proper allowance on the proper ranges. It includes :

- [x] Design a way to properly transition between two different allowances (in terms of speed)
- [x] Implement different ranges rather than just an AllowanceValue
- [x] Modify the allowance added time that will be the target of the binary search

### PR DESCRIPTION
**Pull request by axrolld**
close #603

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceValue.java [hunk 1]
```diff
@@ -22,6 +22,9 @@ public final double getSectionAllowanceTime(
     /** Returns the allowance, given the total time and distance of the trip */
     public abstract double getAllowanceTime(double baseTime, double distance);
 
+    /** Returns the allowance, given the total time and distance of the trip */
+    public abstract double getAllowancePercentage(double baseTime, double distance);
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceValue.java]: Why percentage and not ratio? In my mind percentages are mostly used as a display unit

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java [hunk 2]
```diff
@@ -201,6 +201,12 @@ public double getTotalTime() {
         return ((double) getTotalTimeMS()) / 1000;
     }
 
+    /** Returns the time between two positions of the envelope */
+    public double getTimeBetween(double beginPos, double endPos) {
+        var sliced = make(slice(beginPos, endPos));
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/Envelope.java]: you don't need to  run slice() to get the time between two points, you can get the time at the end, the time at the beginning, and substract. slice makes a copy of the envelope, which is quite heavy just to get the time between two points

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceRange.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package fr.sncf.osrd.envelope_sim.allowances;
+
+/** A range of allowance is a part of the path between [beginPos, endPos] that has a certain allowance value
+ *  Together, allowance ranges are supposed to cover the entire path.
+ *  If a part of the path has no specified value, the default one is used instead. */
+public class AllowanceRange {
+    private final double beginPos;
+    private final double endPos;
+    private final AllowanceValue value;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceRange.java]: you can make this fields public final and remove the getters

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceValue.java (hunk 1 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceValue.java]: I changed it to ratio

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java (hunk 2 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/Envelope.java]: True, that was quite a dumb implementation. I fixed it.

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceRange.java (hunk 3 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AllowanceRange.java]: Done

File: core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStandaloneTrainScheduleParser.java [hunk 4]
```diff
@@ -67,24 +72,66 @@ private static MarecoAllowance parseAllowance(
...
     }
 
+    private static List<AllowanceRange> parseAllowanceRanges(
+            EnvelopePath envelopePath,
+            RJSAllowanceValue defaultValue,
+            RJSAllowanceRange[] ranges
+    ) throws InvalidSchedule {
+        // if no ranges have been defined, just return the default value
+        if (ranges == null || ranges.length == 0) {
+            return List.of(new AllowanceRange(0, envelopePath.getLength(), parseAllowanceValue(defaultValue)));
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStandaloneTrainScheduleParser.java]: can you call parseAllowanceValue(defaultValue) only once, and store the result in a local variable?

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowanceRange.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package fr.sncf.osrd.railjson.schema.schedule;
+
+import com.squareup.moshi.Json;
+
+public final class RJSAllowanceRange {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowanceRange.java:5]: this class lacks a default constructor, which means fields will be initialized to 0 or null
with moshi, all fields are optional. you should initialize these fields to a known default value which is distinguishable from missing, like Double.NaN, and check for that in parseAllowanceRange

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java [hunk 6]
```diff
@@ -99,87 +111,178 @@ private DoubleArrayList findStops(Envelope base) {
...
+                    computeAllowanceRange(envelopeRange, range.value, imposedBeginSpeed, imposedEndSpeed);
+            // memorize the begin and end speeds
+            imposedTransitionSpeeds[rangeIndex] = allowanceRange.getBeginSpeed();
+            imposedTransitionSpeeds[rangeIndex + 1] = allowanceRange.getEndSpeed();
+            res[rangeIndex] = allowanceRange;
+        }
+
+        for (var envelope : res) {
+            builder.addEnvelope(envelope);
+        }
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java]: I think it makes slightly more sense to return res, remove the builder argument, and do that in the caller

**Review by multun — CHANGES_REQUESTED** [review]

**Comment by multun** [comment]
Excellent job, thanks!

File: core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStandaloneTrainScheduleParser.java (hunk 4 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStandaloneTrainScheduleParser.java]: You're right, that's a good point ! I'll do it !

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java (hunk 6 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java]: Done

**Pull request merged**